FLIP-0200: Application of BIP 44 in Flow Wallets (#200)
onflow/flow: flips/20201125-bip-44-multi-account.md

## User Benefit

By following this standard, developers of existing BIP 44 crypto wallets will
be able to reduce friction for users who wish to create accounts on Flow.

## Design Proposal

### Path Levels

BIP 44 defines an implementation of 
[Bitcoin Improvement Proposal 32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki) (BIP 32)
that standardizes a format for hierarchical deterministic (HD) 
key derivation paths that support multiple coins and multiple accounts per coin.

The BIP 44 path format is as follows:

```
m / purpose' / coin_type' / account' / change / address_index
```

When using this format to construct or discover a Flow user wallet, 
the value(s) of each path parameter should follow the specifications below.

#### Purpose

Purpose is always the constant 44' (0x8000002C), 
as described in [BIP 44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#purpose).

#### Coin Type

Coin type is always the constant 539' (0x8000021b), 
the unique identifier for the Flow protocol, as defined in 
[SatoshiLabs Improvement Proposal 44](https://github.com/satoshilabs/slips/blob/master/slip-0044.md) (SLIP 44).

#### Account

> Account is the index of a distinct account address owned by the wallet user. 

Reviewer: This index also has no relation to the `Address state index` used to generate the Flow Addresses, correct? Is that worth noting? It's something i keep thinking about, but am likely an outlier.

Author: Great point. There's no relation, I'll make that more clear

In addition to the above, Flow wallets should follow the standards for account numbering, 
creation and discovery described in [BIP 44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#account).

#### Change

Unlike Bitcoin, Flow does not utilize change addresses. 
Change is always the constant 0.

#### Index

In BIP 44, each index generates a public key that is used to derive
a unique address for an account.
Although Flow does not derive account addresses from public keys,
index can still be used to generate multiple public keys for a single account.

When applying BIP 44 to Flow, each subsequent value for index should be used to
generate a new public key for the same account, starting from 0 and incrementing
by 1.

This index should not be confused with the on-chain account key index.

### Account Discovery

Wallet developers should use a modified version of the original account discovery
procedure [described in BIP 44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#account-discovery).

> 1. Derive the first account's node (index = 0).

Reviewer: I feel like the `node` terminology is first introduced here. May be worth more definition before use? Or is it self explanatory?

Author: "Node" is a term carried over from BIP 44, but good point -- it isn't explained here. I'll add more clarity to this section

2. Scan for addresses by checking each public key index against the public key registry,
respecting the [gap limit described in BIP 44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#address-gap-limit).
3. If no address is found in the registry, stop discovery.
4. If an address is found, query the chain to fetch the account details.
5. If no account is found<sup>1</sup>, skip this index and go to step 1.

> 6. If an account is found, add it to the user's wallet and go to step 1.

Reviewer: Is there an increment step missing? Isn't step 1 going to use index=0 again?

Author: Yep good catch, I'll add!

Implementers of this standard will require a separate source of truth,
the public key registry, to return the associated user address for a given public key. 
This requirement is mentioned in the [related issues](#related-issues) section below.

#### Account Key Index Alignment

The Flow protocol assigns an index to each account key that is registered on chain.
When discussing this proposal, Flow contributors considered a variation that would 
require the path index to always align with the on-chain index for the corresponding
account.

For example, a user may create an account from a seed phrase and generate public keys
at indices 0, 1 and 2. If added to the account in order, their on-chain indices would
also be 0, 1, 2, resulting in a clear mapping from each on-chain key 
to its corresponding derivation path.

The authors of this proposal chose to omit this restriction after considering that,
in many cases, a user may add additional public keys from other sources, thus breaking
the perfect alignment between on-chain and key path indices. Even without outside keys,
it is still possible for keys to be added out of order.

> However, wallet developers or users may still incorrectly assume that the path 
> index of a public key always matches its on-chain account key index.

Reviewer 2: ```suggestion
Wallet developers or users should therefore not rely on the false assumption that the path 
index of a public key always matches its on-chain account key index, as doing so
will cause confusion and errors.
```

Just to spell out the consequences.

Author: This is great, thank you!

#### Potential Misuse of BIP 44

One could argue that this is a misuse of the original BIP 44, BIP 43 and BIP 32
standards based on the fact that Flow does not derive addresses directly 
from cryptographic public keys.

#### Overdependence on BIP 44

This proposal may also promote an overdependence on the BIP 44 standard,
especially for wallets that do not already implement it. 
In some cases it may be better to use a new implementation or
adaptation of BIP 32 that is better suited to the Flow account model.

### Dependencies

* Dependent projects: Flow Ledger application, Flow Port

### Tutorials and Examples

TODO: add examples 

### Compatibility

This proposal does not change any existing APIs. 
The chosen path structure is meant to maximize forward compatibility 
in a variety of use cases.

The convention proposed in this document is intended for use by any wallet
that supports the BIP 44 standard.

In particular, this proposal impacts wallet software that is compatible
with the Flow Ledger application, such as Flow Port.

## Related Issues

### Mapping Public Keys to Addresses

As mentioned in the [drawbacks](#drawbacks) section, this proposal necessitates 

> a public key registry that maps public keys to account addresses.

Reviewer 2: ```suggestion
a public key registry that maps public keys to account addresses. The information required
to associate Flow addresses with their keys is broadcast when keys are added to Flow
accounts both during and after account creation. Making this information searchable
requires indexing this information rather than requiring any additional disclosure on the
part of Flow account owners.
```

be able to reduce friction for users who wish to create accounts on Flow.

## Design Proposal

### Path Levels

BIP 44 defines an implementation of 
[Bitcoin Improvement Proposal 32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki) (BIP 32)
that standardizes a format for hierarchical deterministic (HD) 
key derivation paths that support multiple coins and multiple accounts per coin.

The BIP 44 path format is as follows:

```
m / purpose' / coin_type' / account' / change / address_index
```

When using this format to construct or discover a Flow user wallet, 
the value(s) of each path parameter should follow the specifications below.

#### Purpose

Purpose is always the constant 44' (0x8000002C), 
as described in [BIP 44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#purpose).

#### Coin Type

Coin type is always the constant 539' (0x8000021b), 
the unique identifier for the Flow protocol, as defined in 
[SatoshiLabs Improvement Proposal 44](https://github.com/satoshilabs/slips/blob/master/slip-0044.md) (SLIP 44).

#### Account

Account is the index of a distinct account address owned by the wallet user. 

> No two account indices should be used to generate child public keys that are assigned 

Reviewer 3: ```suggestion
No two distinct account indices 
```
?

any key.

When generating a new key pair for an account, the key index to use in the path to 
generate the new key should be the smallest unused key index available which 
has not yet been used to generate a key on the account.

### Legacy Path

There exist Flow accounts with keys assigned to them which have been generated
by wallets using the legacy path.

The legacy path has been previously used by FCL-Ledger-Web and is what has been used to
generate keys for Ledger Flow users. As far as we are aware, this is the only
application which has used the legacy path.

The legacy path is _deprecated_ by this FLIP, and should not be used to
generate new keys.

The legacy path is:

```
m / 44' / 1' / 769 / 0 / 0
```

All keys generated with the legacy path as of today (Dec 2021) have been done so using the
NIST P-256 elliptic curve.

### Account Discovery

Wallet developers should use a modified version of the original account
discovery procedure [described in BIP
44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#account-discovery).

Modified constants from BIP 44 include:

The key index gap limit should be `5`.

> The account index gap limit should be `5`.

Reviewer 4: Why `5` ? Seems like a lot. Does this mean a wallet should derive 2(two curves) * 5(key_gap) * 5(account_gap) keys in order to say that a wallet is not used? In other words, should the wallet derive 10 keys, just to say the account is not used? Am I getting this wrong? What is the use case for such a big gap limit?

Since keys and accounts should be used in order (meaning the first unused should be picked when creating new account/key) why not gap limit `1`?.

Reviewer 5: Hey @PeterBenc , after consulting with the team, we've concluded that the key gap limit and account gap limit parameters should be rather flexible. Developers who wish to use a smaller key or address gap limit should be able to do so if they choose - developers should use their best judgement when choosing a key gap limit and account gap limit.

Reviewer 4: Thanks for considering this, makes sense, at least for now we are gonna use account_gap_limit 1, and check only the first key (so I guess key_gap_limit 0) since right now, basically nobody is deriving these keys anyway and we can increase the gap limit later if needed. 

Just genuinely interested, I suppose the 5 and 2 (for key account gap respectively) should have some recommending purpose? Since it's supposed to be flexible, what is the reason for defining these?

Reviewer 6: @PeterBenc
In BTC for example addresses are bound to keys, so if you have HD wallet you can have unlimited addresses.

so lets say my wallet pre generates N addresses for me, and I use the Nth to receive some coins. 

After I restore my wallet to another computer, you need to search N addresses (gap = N ) otherwise I see my funds are missing and panic 

i dont know why we need in Flow , I need to reread to FLIP
